### PR TITLE
feat: add data-dawg (my approvals)

### PR DESCRIPTION
**Pull request by w3nda**
Add datadawg.

Using filters I was able to fetch Approval logs for a given address and generate a list of the approvals with token information.

### CODE REVIEW

File: data-dawg/consts/index.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const { utils } = require('ethers');
+
+const BLOCK_LIMIT = 2000;
```
Comment (nikeix) [inline data-dawg/consts/index.js:3]: Unused variable

File: data-dawg/utils/getContractInstanceByAddress.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+const { ethers } = require('ethers');
+const provider = require('./provider');
+
+// creates an ERC20 contract instance by address
+function getContractInstanceByAddress(address) {
```
Comment (nikeix) [inline data-dawg/utils/getContractInstanceByAddress.js:5]: The name isn't describing the functionality very well

File: data-dawg/main.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+          name,
+          amount,
+        })
+      }));
+    } catch (logsError) {
+      console.error(logsError);
+    }
+
+    report.forEach((approval) => {
+      logger('• approval on', approval.name, 'on amount of', approval.amount)
```
Comment (nikeix) [inline data-dawg/main.js:42]: Please print the amount in decimals 👍

File: data-dawg/main.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    try {
+      const logs = await provider.send("eth_getFilterLogs", [filterId]);
+      logger('found', logs.length, 'logs');
+      logger('listing approvals...');
+      logger('-----------------------------------');
+
+      await Promise.all(logs.map(async (transaction) => {
+        const contract = getContractInstanceByAddress(transaction.address);
+        const name = await contract.name();
+        const amount = dataToETH(transaction.data);
```
Comment (nikeix) [inline data-dawg/main.js:30]: Maybe `web3.utils.hexToNumber` is better, check it out

File: data-dawg/utils/normalizeAddress.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// addes 12bytes of zeros to a 20bytes address
+// To match topic requirements.
+function normalizeAddress(address) {
```
Comment (nikeix) [inline data-dawg/utils/normalizeAddress.js:3]: Give it a better name or support any length of address.

Comment (nikeix) [inline data-dawg/utils/normalizeAddress.js:3]: Or use https://web3js.readthedocs.io/en/v1.2.0/web3-utils.html#padleft

File: data-dawg/utils/getContractInstanceByAddress.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  const abi = [
```
Comment (nikeix) [inline data-dawg/utils/getContractInstanceByAddress.js:6]: I would use the original [ERC20 abi](https://gist.github.com/veox/8800debbf56e24718f9f483e1e40c35c) for future uses. You might need decimals soon 💵

File: data-dawg/main.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    try {
+      const logs = await provider.send("eth_getFilterLogs", [filterId]);
+      logger('found', logs.length, 'logs');
+      logger('listing approvals...');
+      logger('-----------------------------------');
+
+      await Promise.all(logs.map(async (transaction) => {
+        const contract = getContractInstanceByAddress(transaction.address);
+        const name = await contract.name();
```
Comment (nikeix) [inline data-dawg/main.js:29]: Do some caching 💵
Calling a contract function every time is not efficient

File: data-dawg/main.js (hunk 3 above)
Comment (nikeix) [reply data-dawg/main.js:42]: e.g `1.8 eth` instead of `180000000000000`

**Review by nikeix — CHANGES_REQUESTED** [review]

File: data-dawg/utils/getContractInstanceByAddress.js (hunk 2 above)
Comment (w3nda) [reply data-dawg/utils/getContractInstanceByAddress.js:5]: You are right.
Changed to `getERC20ContractInstanceByAddress`

File: data-dawg/utils/normalizeAddress.js (hunk 5 above)
Comment (w3nda) [reply data-dawg/utils/normalizeAddress.js:3]: Used `ethers.utils.hexZeroPad` (equivalent)

File: data-dawg/main.js (hunk 4 above)
Comment (w3nda) [reply data-dawg/main.js:30]: Both relay on `parseInt`. doesn't really matter.

Comment (nikeix) [reply data-dawg/main.js:30]: It's extra code. `dataToETH` is redundant

**Pull request merged**